java/ql/src/semmle/code/java/dataflow/internal/DataFlowImplCommon.qll

```diff
@@ -1,915 +1,866 @@
       ) {
-        exists(ParameterNode param, CallContext innercc |
-          viableParamArg(call, param, arg, cc, innercc) and
-          parameterValueFlowsThrough(param, kind, innercc)
+        exists(ParameterNode param | viableParamArg(call, param, arg) |
+          parameterValueFlowReturn(param, _, kind, contentIn, contentOut)
         )
       }
 
       /**
        * Holds if `arg` flows to `out` through a call using only value-preserving steps,
-       * in call context cc.
+       * not taking call contexts into account.
+       *
+       * `contentIn` describes the content of `arg` that can flow to `out` (if any), and
+       * `contentOut` describes the content of `out` that it flows to (if any).
        */
-      cached
-      predicate argumentValueFlowsThrough(ArgumentNode arg, OutNode out, CallContext cc) {
-        exists(DataFlowCall call, ReturnKind kind |
-          argumentValueFlowsThrough0(call, arg, kind, cc)
-        |
-          out = getAnOutNode(call, kind) and
-          not isUnreachableInCall(out, cc.(CallContextSpecificCall).getCall()) and
-          compatibleTypes(getErasedNodeType(arg), getErasedNodeType(out))
+      private predicate argumentValueFlowsThrough(
+        ArgumentNode arg, Node out, ContentOption contentIn, ContentOption contentOut
+      ) {
+        exists(DataFlowCall call, ReturnKindExt kind |
+          argumentValueFlowsThrough0(call, arg, kind, contentIn, contentOut) and
+          out = kind.getAnOutNode(call)
         )
       }
     }
 
-    /**
-     * Holds if the call context `call` either improves virtual dispatch in
-     * `callable` or if it allows us to prune unreachable nodes in `callable`.
-     */
-    cached
-    predicate recordDataFlowCallSite(DataFlowCall call, DataFlowCallable callable) {
-      reducedViableImplInCallContext(_, callable, call)
-      or
-      exists(Node n | n.getEnclosingCallable() = callable | isUnreachableInCall(n, call))
-    }
-
-    cached
-    newtype TCallContext =
-      TAnyCallContext() or
-      TSpecificCall(DataFlowCall call) { recordDataFlowCallSite(call, _) } or
-      TSomeCall() or
-      TReturn(DataFlowCallable c, DataFlowCall call) { reducedViableImplInReturn(c, call) }
-
-    cached
-    newtype TReturnPosition =
-      TReturnPosition0(DataFlowCallable c, ReturnKindExt kind) { returnPosition(_, c, kind) }
-
-    cached
-    newtype TLocalFlowCallContext =
-      TAnyLocalCall() or
-      TSpecificLocalCall(DataFlowCall call) { isUnreachableInCall(_, call) }
+    import Final
   }
 
-  pragma[noinline]
-  private predicate returnPosition(ReturnNodeExt ret, DataFlowCallable c, ReturnKindExt kind) {
-    c = returnNodeGetEnclosingCallable(ret) and
-    kind = ret.getKind()
+  /**
+   * Holds if `p` can flow to the pre-update node associated with post-update
+   * node `n`, in the same callable, using only value-preserving steps.
+   */
+  cached
+  predicate parameterValueFlowsToPreUpdate(ParameterNode p, PostUpdateNode n) {
```

Comment:
The cached predicates that are part of the SCC should be inside the `Final` module, such that the SCC is kept together.